### PR TITLE
Full core structure developed

### PR DESCRIPTION
**Pull request by chemphys**
The main infrastructure of the software is implemented and working:
- Electrostatics has been added. Works.
- Ion water potential (for now, only fluoride) has been added. Works.

@dgasmith , there are a couple of formatting issues that you mentioned that still need to be fixed. I will do it before merging.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -92,7 +92,7 @@ install:
 # END MRR 
   - cmake -H. -Bbuild
   - cd build
-  - travis_wait 40 make
+  - travis_wait 60 make
```
Comment (dgasmith) [inline .travis.yml]: Kind of surprised this is taking as long as it does. One strategy is pull out as much of the C++ headers from the polynomial files as you can. You C can compile 1-2 orders of magnitude faster than C++. It seems a bit silly, but I noted its much more enjoyable to work on things that dont require half an hour to compile.

File: src/main/CMakeLists.txt [hunk 2]
```diff
@@ -4,11 +4,12 @@ target_include_directories(clusters_ultimate PRIVATE ${CMAKE_SOURCE_DIR}/../exte
 
 target_link_libraries(clusters_ultimate bblock)
 target_link_libraries(clusters_ultimate io_tools)
-target_link_libraries(clusters_ultimate tools)
+#target_link_libraries(clusters_ultimate tools)
 target_link_libraries(clusters_ultimate 1b)
 target_link_libraries(clusters_ultimate 2b)
 target_link_libraries(clusters_ultimate 3b)
 target_link_libraries(clusters_ultimate dispersion)
+target_link_libraries(clusters_ultimate electrostatics)
```
Comment (dgasmith) [inline src/main/CMakeLists.txt]: Glad the CMake overhaul looks like its working out.

File: src/potential/2b/energy2b.cpp [hunk 3]
```diff
@@ -4,15 +4,25 @@ namespace e2b {
 
 double get_2b_energy(std::string m1, std::string m2, size_t nm,
                      std::vector<double> xyz1, std::vector<double> xyz2) {
+
+  // Order the two monomer names and corresponding xyz
   if (m2 < m1) {
     std::string tmp = m1;
     m1 = m2;
     m2 = tmp;
+    std::vector<double> tmp2 = xyz1;
```
Comment (dgasmith) [inline src/potential/2b/energy2b.cpp]: You might want to use a `std::move` here and below depending on your requirements.

File: src/potential/2b/energy2b.cpp [hunk 4]
```diff
@@ -22,19 +32,44 @@ double get_2b_energy(std::string m1, std::string m2, size_t nm,
 double get_2b_energy(std::string m1, std::string m2,size_t nm, 
                      std::vector<double> xyz1, std::vector<double> xyz2,
                      std::vector<double> &grd1, std::vector<double> &grd2) {
+  // Order the two monomer names and corresponding xyz
+  bool swaped = false;
   if (m2 < m1) {
     std::string tmp = m1;
     m1 = m2;
     m2 = tmp;
+    std::vector<double> tmp2 = xyz1;
```
Comment (dgasmith) [inline src/potential/2b/energy2b.cpp]: Same `std::move` comment. It should be doing the exact same thing, but always nice to be verbose about these things IMO.

File: src/potential/electrostatics/gammq.cpp [hunk 5]
```diff
@@ -0,0 +1,181 @@
+#include <cmath>
```
Comment (dgasmith) [inline src/potential/electrostatics/gammq.cpp]: Where did you pull this from? They might require their license at the top of this file.

**Comment by dgasmith** [comment]
Overall, I think is coming together quite well. I think once we get to the profiling (VTune) step it will become quite apparent where we need to focus on. As a note you will need to up your testing scripts to either Python or C++ due to the tolerance issues.

File: src/potential/electrostatics/gammq.cpp (hunk 5 above)
Comment (chemphys) [reply src/potential/electrostatics/gammq.cpp]: This is code from numerical recipes. It was in our old c++ software, but no reference from where was it taken. i found out it was from NR when I was looking for a vectorizable form of obtaining gamma functions. I will add a comment with probably the citation to the book.

File: .travis.yml (hunk 1 above)
Comment (chemphys) [reply .travis.yml]: What do you mean? Comile the polynomial files using gcc/icc instead of g++/icpc?

Comment (dgasmith) [reply .travis.yml]: No, simply remove as many C++ headers from being included into the kernel files as possible. If its not possible, its not a big deal. Just a suggestion.

Comment (chemphys) [reply .travis.yml]: I will see if I can remove some of them.

File: src/potential/electrostatics/gammq.cpp (hunk 5 above)
Comment (chemphys) [reply src/potential/electrostatics/gammq.cpp]: Apparently the functions are taken from [here](https://www2.units.it/ipl/students_area/imm2/files/Numerical_Recipes.pdf). They say that it has "limited use". Have you ever used a routine from here? Can we use it? The functions that vlodimir (a former postdoc in our lab) added are slightly modified.

Comment (dgasmith) [reply src/potential/electrostatics/gammq.cpp]: If its from the book I would cite the book and add a mention that the post-doc modified them.

Comment (chemphys) [reply src/potential/electrostatics/gammq.cpp]: Will do that.

**Comment by chemphys** [comment]
I will merge this pull request and start playing with VTune. 
I will keep you posted!

**Pull request merged**